File: server/app/db/models/catalog.py
"""Categories, packing templates and the shopping list."""

from __future__ import annotations

import uuid

from sqlalchemy import Boolean, Index, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PgUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, HouseholdScopedMixin, sync_fk


class Category(Base, HouseholdScopedMixin):
    __tablename__ = "categories"

    name: Mapped[str] = mapped_column(Text, nullable=False)
    color_hex: Mapped[str] = mapped_column(String(9), nullable=False, server_default="#8E8E93")
    icon: Mapped[str | None] = mapped_column(Text, nullable=True)

    __table_args__ = (Index("ix_categories_household_id_updated_at", "household_id", "updated_at"),)


class PackingTemplate(Base, HouseholdScopedMixin):
    """A named list of things to bring; applying it creates subtasks."""

    __tablename__ = "packing_templates"

    name: Mapped[str] = mapped_column(Text, nullable=False)
    # A JSON array of strings. The items have no identity of their own — the
    # template is edited as a whole, so last-write-wins applies to the list.
    items: Mapped[list] = mapped_column(JSONB, nullable=False, server_default=text("'[]'::jsonb"))

    __table_args__ = (
        Index("ix_packing_templates_household_id_updated_at", "household_id", "updated_at"),
    )


class ShoppingItem(Base, HouseholdScopedMixin):
    """An entity of its own, not a task."""

    __tablename__ = "shopping_items"

    title: Mapped[str] = mapped_column(Text, nullable=False)
    # Free text rather than a number: "2", "2 kg" and "a pack" are all typed
    # into the same field in practice.
    quantity: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_bought: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("false"))
    category_id: Mapped[uuid.UUID | None] = mapped_column(
        PgUUID(as_uuid=True), sync_fk("categories.id"), nullable=True
    )
    added_by: Mapped[uuid.UUID] = mapped_column(
        PgUUID(as_uuid=True), sync_fk("users.id"), nullable=False
    )

    __table_args__ = (
        Index("ix_shopping_items_household_id_updated_at", "household_id", "updated_at"),
        Index("ix_shopping_items_category_id", "category_id"),
    )

File: ios/FamilyCalendarKit/Sources/FamilyCalendarKit/Models/Catalog.swift
import Foundation
import GRDB

/// Named `TaskCategory`, not `Category`: the plain name is ambiguous the moment
/// a file imports SwiftUI or UIKit alongside this module, and every view that
/// shows a colour needs both. The same reasoning as `CalendarTask`. The table is
/// still `categories`.
public struct TaskCategory: HouseholdScopedRecord {
    public var id: UUID
    public var householdID: UUID
    public var name: String
    public var colorHex: String
    public var icon: String?

    public var createdAt: Date
    public var updatedAt: Date
    public var updatedBy: UUID?
    public var deletedAt: Date?
    public var dirty: Bool

    public init(
        id: UUID = UUID(),
        householdID: UUID,
        name: String,
        colorHex: String = "#8E8E93",
        icon: String? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        updatedBy: UUID? = nil,
        deletedAt: Date? = nil,
        dirty: Bool = true
    ) {
        self.id = id
        self.householdID = householdID
        self.name = name
        self.colorHex = colorHex
        self.icon = icon
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.updatedBy = updatedBy
        self.deletedAt = deletedAt
        self.dirty = dirty
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case householdID = "household_id"
        case name
        case colorHex = "color_hex"
        case icon
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case updatedBy = "updated_by"
        case deletedAt = "deleted_at"
        case dirty
    }
}

extension TaskCategory: TableRecord {
    public static let databaseTableName = "categories"

    public enum Columns {
        public static let id = Column(CodingKeys.id)
        public static let householdID = Column(CodingKeys.householdID)
        public static let name = Column(CodingKeys.name)
        public static let updatedAt = Column(CodingKeys.updatedAt)
        public static let deletedAt = Column(CodingKeys.deletedAt)
        public static let dirty = Column(CodingKeys.dirty)
    }
}

/// A named list of things to bring. Applying it creates subtasks in one tap.
public struct PackingTemplate: HouseholdScopedRecord {
    public var id: UUID
    public var householdID: UUID
    public var name: String
    /// JSON array of strings; the items have no identity of their own.
    public var items: [String]

    public var createdAt: Date
    public var updatedAt: Date
    public var updatedBy: UUID?
    public var deletedAt: Date?
    public var dirty: Bool

    public init(
        id: UUID = UUID(),
        householdID: UUID,
        name: String,
        items: [String] = [],
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        updatedBy: UUID? = nil,
        deletedAt: Date? = nil,
        dirty: Bool = true
    ) {
        self.id = id
        self.householdID = householdID
        self.name = name
        self.items = items
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.updatedBy = updatedBy
        self.deletedAt = deletedAt
        self.dirty = dirty
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case householdID = "household_id"
        case name
        case items
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case updatedBy = "updated_by"
        case deletedAt = "deleted_at"
        case dirty
    }
}

extension PackingTemplate: TableRecord {
    public static let databaseTableName = "packing_templates"

    public enum Columns {
        public static let id = Column(CodingKeys.id)
        public static let householdID = Column(CodingKeys.householdID)
        public static let name = Column(CodingKeys.name)
        public static let updatedAt = Column(CodingKeys.updatedAt)
        public static let deletedAt = Column(CodingKeys.deletedAt)
        public static let dirty = Column(CodingKeys.dirty)
    }
}

/// An entity of its own, not a task.
public struct ShoppingItem: HouseholdScopedRecord {
    public var id: UUID
    public var householdID: UUID
    public var title: String
    /// Free text: "2", "2 kg" and "a pack" all get typed into this field.
    public var quantity: String?
    public var isBought: Bool
    public var categoryID: UUID?
    public var addedBy: UUID

    public var createdAt: Date
    public var updatedAt: Date
    public var updatedBy: UUID?
    public var deletedAt: Date?
    public var dirty: Bool

    public init(
        id: UUID = UUID(),
        householdID: UUID,
        title: String,
        quantity: String? = nil,
        isBought: Bool = false,
        categoryID: UUID? = nil,
        addedBy: UUID,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        updatedBy: UUID? = nil,
        deletedAt: Date? = nil,
        dirty: Bool = true
    ) {
        self.id = id
        self.householdID = householdID
        self.title = title
        self.quantity = quantity
        self.isBought = isBought
        self.categoryID = categoryID
        self.addedBy = addedBy
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.updatedBy = updatedBy
        self.deletedAt = deletedAt
        self.dirty = dirty
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case householdID = "household_id"
        case title
        case quantity
        case isBought = "is_bought"
        case categoryID = "category_id"
        case addedBy = "added_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case updatedBy = "updated_by"
        case deletedAt = "deleted_at"
        case dirty
    }
}

extension ShoppingItem: TableRecord {
    public static let databaseTableName = "shopping_items"

    public enum Columns {
        public static let id = Column(CodingKeys.id)
        public static let householdID = Column(CodingKeys.householdID)
        public static let isBought = Column(CodingKeys.isBought)
        public static let updatedAt = Column(CodingKeys.updatedAt)
        public static let deletedAt = Column(CodingKeys.deletedAt)
        public static let dirty = Column(CodingKeys.dirty)
    }
}
